File: frontend/src/features/admin/lib/saturdayWeek.ts
const SATURDAY = 6; // JS getUTCDay(): Sun=0..Sat=6

function utcDate(year: number, month: number, day: number): Date {
  return new Date(Date.UTC(year, month, day));
}

export function firstSaturday(year: number): Date {
  const jan1 = utcDate(year, 0, 1);
  const delta = (SATURDAY - jan1.getUTCDay() + 7) % 7;
  return utcDate(year, 0, 1 + delta);
}

function lastSaturdayOnOrBefore(d: Date): Date {
  const delta = (d.getUTCDay() - SATURDAY + 7) % 7;
  return new Date(d.getTime() - delta * 86_400_000);
}

export function weeksInYear(year: number): number {
  const start = firstSaturday(year);
  const end = lastSaturdayOnOrBefore(utcDate(year, 11, 31));
  return Math.floor((end.getTime() - start.getTime()) / (7 * 86_400_000)) + 1;
}

export function saturdayWeekRange(year: number, week: number): [Date, Date] {
  const max = weeksInYear(year);
  if (week < 1 || week > max) {
    throw new RangeError(`week ${week} out of range for year ${year} (1..${max})`);
  }
  const start = new Date(firstSaturday(year).getTime() + (week - 1) * 7 * 86_400_000);
  const end = new Date(start.getTime() + 6 * 86_400_000);
  return [start, end];
}

export function weekOfDate(d: Date): [number, number] {
  const saturday = lastSaturdayOnOrBefore(d);
  const year = saturday.getUTCFullYear();
  const fs = firstSaturday(year);
  if (saturday.getTime() < fs.getTime()) {
    const prev = year - 1;
    const prevFs = firstSaturday(prev);
    const week = Math.floor((saturday.getTime() - prevFs.getTime()) / (7 * 86_400_000)) + 1;
    return [prev, week];
  }
  const week = Math.floor((saturday.getTime() - fs.getTime()) / (7 * 86_400_000)) + 1;
  return [year, week];
}

File: src/collector/saturday_week.py
"""Saturday-anchored week math.

Convention used by the admin Beatport ingest UI.

- Week N of year Y starts on Saturday(Y, N) and ends on the following Friday.
- Saturday(Y, 1) is the first Saturday on or after Jan 1 of Y.
- Days from Jan 1 up to (but excluding) Saturday(Y, 1) belong to the last
  week of Y - 1.
"""

from __future__ import annotations

from datetime import date, timedelta

_SATURDAY = 5  # date.weekday(): Mon=0 .. Sun=6


def first_saturday(year: int) -> date:
    jan1 = date(year, 1, 1)
    delta = (_SATURDAY - jan1.weekday()) % 7
    return jan1 + timedelta(days=delta)


def _last_saturday_on_or_before(d: date) -> date:
    delta = (d.weekday() - _SATURDAY) % 7
    return d - timedelta(days=delta)


def weeks_in_year(year: int) -> int:
    start = first_saturday(year)
    end = _last_saturday_on_or_before(date(year, 12, 31))
    return ((end - start).days // 7) + 1


def saturday_week_range(year: int, week: int) -> tuple[date, date]:
    limit = weeks_in_year(year)
    if week < 1 or week > limit:
        raise ValueError(
            f"week {week} out of range for year {year} (1..{limit})"
        )
    start = first_saturday(year) + timedelta(days=(week - 1) * 7)
    end = start + timedelta(days=6)
    return start, end


def week_of_date(d: date) -> tuple[int, int]:
    saturday = _last_saturday_on_or_before(d)
    year = saturday.year
    fs = first_saturday(year)
    if saturday < fs:
        # Saturday belongs to previous year's last week.
        prev = year - 1
        prev_fs = first_saturday(prev)
        week = ((saturday - prev_fs).days // 7) + 1
        return prev, week
    week = ((saturday - fs).days // 7) + 1
    return year, week
